File: day1/part-two.js
/*
 * More standard javascript, using 'some' as an iterator so that I can
 * break when the first basement instruction is found.
 *
 * This solution doesn't handle newlines like the python version, if I
 * needed multi-line support I would just count the newlines as I was
 * going and then subtract them from the index at the end
 */

'use strict';
const fs = require('fs');

let fileName = process.argv[2] || 'input.txt';
fs.readFile(fileName, (err, data) => {
  console.log(getFirstBasementIndex(String(data)));
});

function getFirstBasementIndex(directions) {
  let currentFloor = 0;
  let firstBasement;

  directions.split('').some((floor, index) => {
    if (floor === '(') {
      currentFloor++;
    } else if (floor === ')') {
      currentFloor--;
    }

    if (currentFloor === -1) {
      firstBasement = index + 1;
      return true;
    }
  });

  return firstBasement;
}

File: day1/part-two.py
#
# Learnt about enumerate for when index tracking is required (rather than just
# keeping track of it yourself, apparently this way is more pythonic). Decided
# to support multi-line files just in case.
#

import sys

def getFirstBasementIndex(directionsFile):
    currentFloor = 0
    firstBasement = 0
    totalIndex = 0
    for line in directionsFile:
        for index, floor in enumerate(line):
            if floor == '(':
                currentFloor += 1
            elif floor == ')':
                currentFloor -= 1

            if currentFloor == -1:
                firstBasement = totalIndex + index + 1
                return firstBasement
        # Minus 1 for newline characters
        totalIndex += len(line) - 1

    return None

if len(sys.argv) > 1:
    fileName = sys.argv[1]
else:
    fileName = 'input.txt'
directionsFile = open(fileName, 'r')
print getFirstBasementIndex(directionsFile)
directionsFile.close()
